javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java

```diff
@@ -160,7 +140,69 @@ public static ResettableClassFileTransformer installBytebuddyAgent(
 
     agentBuilder = customizeByteBuddyAgent(agentBuilder);
     log.debug("Installed {} instrumenter(s)", numInstrumenters);
-    return agentBuilder.installOn(inst);
+    ResettableClassFileTransformer resettableClassFileTransformer = agentBuilder.installOn(inst);
+    installComponentsAfterByteBuddy(componentInstallers);
+    return resettableClassFileTransformer;
+  }
+
+  private static void installComponentsBeforeByteBuddy(
+      Iterable<ComponentInstaller> componentInstallers) {
+    ClassLoader savedContextClassLoader = Thread.currentThread().getContextClassLoader();
+    try {
+      // calling (shaded) OpenTelemetry.getGlobalTracerProvider() with context class loader set to
+      // the
+      // agent class loader, so that SPI finds the agent's (isolated) SDK, and (shaded)
+      // OpenTelemetry registers it, and then when instrumentation calls (shaded)
+      // OpenTelemetry.getGlobalTracerProvider() later, they get back the agent's (isolated) SDK
+      //
+      // but if we don't trigger this early registration, then if instrumentation is the first to
+      // call (shaded) OpenTelemetry.getGlobalTracerProvider(), then SPI can't see the agent class
+      // loader,
+      // and so (shaded) OpenTelemetry registers the no-op TracerFactory, and it cannot be replaced
+      // later
+      Thread.currentThread().setContextClassLoader(AgentInstaller.class.getClassLoader());
+      for (ComponentInstaller componentInstaller : componentInstallers) {
+        componentInstaller.beforeByteBuddyAgent();
+      }
+    } finally {
+      Thread.currentThread().setContextClassLoader(savedContextClassLoader);
+    }
+  }
+
+  private static void installComponentsAfterByteBuddy(
+      Iterable<ComponentInstaller> componentInstallers) {
+    /*
+     * java.util.logging.LogManager maintains a final static LogManager, which is created during class initialization.
+     *
+     * JMXFetch uses jre bootstrap classes which touch this class. This means applications which require a custom log
+     * manager may not have a chance to set the global log manager if jmxfetch runs first. JMXFetch will incorrectly
+     * set the global log manager in cases where the app sets the log manager system property or when the log manager
+     * class is not on the system classpath.
+     *
+     * Our solution is to delay the initialization of jmxfetch when we detect a custom log manager being used.
+     *
+     * Once we see the LogManager class loading, it's safe to start jmxfetch because the application is already setting
+     * the global log manager and jmxfetch won't be able to touch it due to classloader locking.
+     */
+
+    /*
+     * Similar thing happens with AgentTracer on (at least) zulu-8 because it uses OkHttp which indirectly loads JFR
+     * events which in turn loads LogManager. This is not a problem on newer JDKs because there JFR uses different
+     * logging facility.
+     */
```

Comment:
JMXFetch is a DataDog thing that's no longer present in OTel javaagent, we should consider removing this code in separate issue/PR.

Reply:
I can clean it up once merged

Reply:
Wait, is this delay needed at all now? We don't have `JMXFetch`, `AgentTracer` certainly does not use `OkHttp`. Do we need this?

And if yes, then do we need this delay here, for all components, or just inside `OpenTelemetryInstaller`?

Reply:
I believe we still need the delay if we are using an exporter that uses SSL.

This bit me also when 8u231 was released, at the time I was convinced this enhancement was the cause: https://bugs.openjdk.java.net/browse/JDK-8148188

I think the chain goes like:

Agent -> SSL -> JFR -> j.u.l.

and then thankfully in Java 9, JFR doesn't use j.u.l. anymore, which prevents this chain from reaching j.u.l.

Reply:
@pavolloffay @trask  then do we need this delay here, for all components, or just inside `OpenTelemetryInstaller`?

Reply:
seems ok to do it for all, so the components themselves don't need to worry about this weirdness